File: Sources/NotchAgent/Core/Utilities/AppPaths.swift
import Foundation

enum AppPaths {
    static var home: URL {
        FileManager.default.homeDirectoryForCurrentUser
    }

    /// ~/Library/Application Support/NotchAgent (created on first access).
    static var appSupport: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("NotchAgent", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }
}

struct FileStamp: Equatable, Sendable {
    let modified: Date
    let size: Int

    init?(url: URL) {
        // FileManager instead of URL.resourceValues: NSURL caches resource
        // values per instance, which would freeze the stamp of the live,
        // constantly-growing transcript within a refresh cycle.
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
              let modified = attributes[.modificationDate] as? Date,
              let size = (attributes[.size] as? NSNumber)?.intValue
        else { return nil }
        self.modified = modified
        self.size = size
    }
}

/// Recursively lists files with `ext`, modified after `cutoff`, newest first, capped.
func recentFiles(under root: URL, ext: String, modifiedAfter cutoff: Date, limit: Int = 500) -> [URL] {
    let keys: [URLResourceKey] = [.contentModificationDateKey, .isRegularFileKey]
    guard let enumerator = FileManager.default.enumerator(
        at: root, includingPropertiesForKeys: keys, options: [.skipsHiddenFiles]
    ) else { return [] }

    var result: [(URL, Date)] = []
    for case let url as URL in enumerator {
        guard url.pathExtension == ext,
              let values = try? url.resourceValues(forKeys: Set(keys)),
              values.isRegularFile == true,
              let modified = values.contentModificationDate,
              modified >= cutoff
        else { continue }
        result.append((url, modified))
    }
    return result
        .sorted { $0.1 > $1.1 }
        .prefix(limit)
        .map(\.0)
}

File: windows/NotchAgent.Windows/Providers/Shared/AppPaths.cs
namespace NotchAgent.Windows.Providers.Shared;

public static class AppPaths
{
    public static string Home => Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);

    /// %APPDATA%\NotchAgent (created on first access).
    public static string AppSupport
    {
        get
        {
            var dir = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "NotchAgent");
            Directory.CreateDirectory(dir);
            return dir;
        }
    }
}

public readonly record struct FileStamp(DateTimeOffset Modified, long Size)
{
    public static FileStamp? Of(string path)
    {
        try
        {
            var info = new FileInfo(path);
            if (!info.Exists) return null;
            return new FileStamp(info.LastWriteTimeUtc, info.Length);
        }
        catch (IOException)
        {
            return null;
        }
    }
}

public static class FileScan
{
    /// Recursively lists files with `ext`, modified after `cutoff`, newest first, capped.
    public static List<string> RecentFiles(string root, string ext, DateTimeOffset cutoff, int limit = 500)
    {
        if (!Directory.Exists(root)) return new List<string>();
        var results = new List<(string Path, DateTimeOffset Modified)>();
        foreach (var file in Directory.EnumerateFiles(root, $"*.{ext}", SearchOption.AllDirectories))
        {
            DateTimeOffset modified;
            try { modified = File.GetLastWriteTimeUtc(file); }
            catch (IOException) { continue; }
            if (modified >= cutoff)
            {
                results.Add((file, modified));
            }
        }
        return results
            .OrderByDescending(r => r.Modified)
            .Take(limit)
            .Select(r => r.Path)
            .ToList();
    }
}
